Update section 2 with several additions for review (#52)
j3-fortran/fpp-202y: drafts/25-xxx-specifications.txt

handling, and tokenization for the preprocessor.)

2 Lexical specifications
------------------------

2.1 Lines
---------

li00. The Fortran preprocessor recognizes three distinct types of lines:
      preprocessing directives (and continuation lines thereof),
      Fortran comment lines, and Fortran source fragments.

li11. A line that has a '#'
      character as the first non-blank character of the line is a

>       directive line (as required by C 2023 6.10.1 paragraph 2),

Reviewer: We are using `C 2023` throughout to refer to the C spec, but the listing in the references is:

```
ISO/IEC 9899:2023 Programming languages -- C
                 (working draft N3096)
```

I see two options:
1. Replace all citations with `ISO/IEC 9899:2023`, or
2. Change the references to introduce a "short form", perhaps something like:

```
ISO/IEC 9899:2023 Programming languages -- C ("C 2023")
                 (working draft N3096)
```

Author: I'll update this PR.

Author: OK, I think I have updated the PR, but I am easily confused.

      except when otherwise specified by the next two rules.

li13. In fixed-form input files, a '#' in column 6 of a non-comment
      line does not introduce a directive line.

li15. A preprocessor directive can be continued with a backslash '\'
      immediately followed by a new-line. The backslash and new-line are
      deleted, the content of the subsequent line are textually appended
      to the current directive, and the subsequent line is deleted.
      This process repeats until the current directive does not end
      with a backslash '\' immediately followed by a new-line.

li17. Preprocessor directive continuation processing described by the prior rule is
      effectively performed before any other processing of the text in affected lines.

li19. The maximum length of a preprocessor directive (including
      continuation text) is one million characters.

li21. A non-empty source file shall not end with a directive line
      that ends with a '\' followed
      immediately by a new-line (as required in C 2023 5.1.1.2 bullet 2).

li31. Fortran comment lines are defined as in 25-007 6.2.2.3 and

>       6.2.3.2.

Reviewer: The referenced sections don't exist.

I think you mean 6.3.2.3 and 6.3.3.2

Author: I believe the technical description for this error is "dammit!"

I don't know how to suggest my own changes to the PR, so feel free to suggest the correction and I will be happy to commit it.

If not, I can update the PR. (I want to fix all your comments...)

Reviewer: ```suggestion
li31. Fortran comment lines are defined as in 25-007 6.3.2.3 and
      6.3.3.2.
```

co05. '/*' ... '*/' comments on directive lines shall extend past a
      new-line only if the line ends in '\' new-line, indicating a
      continuation line.

co07. '/*' ... '*/' comments on directive lines are replaced by a
      single space, as specified in C 2023 5.1.1.2 bullet 3.

co09. In a directive line, the '!' character is not interpreted as introducing
      a Fortran-style comment, and neither the `!` character nor any subsequent
      text are removed by the preprocessor.

co11. Directive lines (by definition) cannot contain Fortran
      fixed-form 'C' or '*' style comments.

2.5 Token lexicon
-----------------

The preprocessor decomposes the source file into preprocessing tokens (see C 2023 5.1.1.2
Translation phases).
As such, there is a specific lexicon of tokens recognized by
the preprocessor (including unrecognizable tokens).

We use illustrative syntax to describe the directive specifications,
and the translation behavior of the preprocess on Fortran comment lines
and Fortran source fragment lines.

This illustrative syntax makes use of these "tokens".
Further definition of the recognized tokens is deferred to the upcoming preprocessor syntax paper.

> to01. In the definitions of object macros and function-like macros,
>       the tokens that may appear in a replacement list include those
>       allowed in Fortran source, those allowed in C integer expressions,
>       and any additional tokens recognized by the processor.

Reviewer: Given that "any additional tokens recognized by the processor" includes "unrecognized token", 
this rule really boils down to approximately mean: "the replacement list may include any arbitrary sequence of characters that doesn't include a new-line". 

I think it would be clearer to just state that directly, how about:

```suggestion
to01. In the definitions of object macros and function-like macros,
      the replacement list may include any arbitrary sequence of 
      characters that doesn't include a new-line. Once tokenized,
      this for example may include any tokens allowed in Fortran 
      source lines, those allowed in C integer expressions,
      and any additional tokens recognized by the processor.
```

to03. Without naming all the tokens explicitly, they appear in the illustrative
      syntax in subsequent sections.

> to07. The '#' character on a directive line may be preceded or followed
>       by whitespace.

Reviewer: This is ambiguously worded, and I believe it's now entirely duplicative of ws02

```suggestion
```

------------------------------

ws01. The whitespace characters blank and horizontal tab character may appear
      on directive lines.

ws03. Whitespace characters are significant in determining token
      boundaries in preprocessor directive lines.

ws05. Outside of character constants, multiple whitespace characters
      on a directive line are treated as a single space.

ws07. Whitespace characters are significant in determining token
      boundaries for the purposes of recognizing macro names,
      in both fixed-form and free-form Fortran source fragments.

ws09. Whitespace characters are significant in determining token
      boundaries for the purposes of recognizing macro names,
      in both fixed-form and free-form Fortran comment lines.

ws11. In fixed-form input, macro names are not recognized as such when
      spaces are inserted into their invocations.

2.4 Comments
------------

co01. Directive lines may contain C-style '/*' ... '*/' comments.

co03. Directive lines shall not contain C style '//' comments.

co05. '/*' ... '*/' comments on directive lines shall extend past a
      new-line only if the line ends in '\' new-line, indicating a

>       continuation line. (Maybe controversial)

Reviewer: I think we can remove the non-normative commentary here.

Everything is potentially controversial.

```suggestion
      continuation line.
```

co07. '/*' ... '*/' comments on directive lines are replaced by a
      single space, as specified in C 2023 5.1.1.2 bullet 3.

co09. In a directive line, the '!' character is not interpreted as introducing
      a Fortran-style comment, and neither the `!` character nor any subsequent
      text are removed by the preprocessor.

co11. Directive lines (by definition) cannot contain Fortran
      fixed-form 'C' or '*' style comments.

2.5 Token lexicon
-----------------

The preprocessor is a token-manipulation engine (See C 2023 5.1.1.2
Translation phases).
As such, there is a specific lexicon of tokens recognized by
the preprocessor (including unrecognizable tokens).

We use illustrative syntax to describe the directive specifications,

> and the execution behavior of the preprocess on Fortran comment lines

Reviewer: I continue to strongly oppose misusing the term "execution" to refer to any part of compilation.

I believe the C spec exclusively uses the term "execution" to describe program execution. Any actions taken by the preprocessor or compiler are "translation" (see C 23 5.0-1). Based on a quick skim, the same appears to be true of the Fortran spec.

```suggestion
and the translation behavior of the preprocess on Fortran comment lines
```

and Fortran source fragment lines.

This illustrative syntax makes use of these "tokens".
Further definition of the recognized tokens is deferred to the upcoming preprocessor syntax paper.

to01. In definitions of object macros and function/like macros,
      the tokens that may appear in a replacement list include those

>       allowed in Fortran and those allowed in C integer expressions.

Reviewer: As written, this rule could easily be misread as overly constraining. 

In CPP, a `#define` replacement list may contain literally any text (other than a newline). This includes things like C keywords, or even unknown tokens (e.g. `#$&%`).

The rule is technically accurate because indeed the "tokens .. may .. include" the subset you name, but they also may include plenty of other tokens outside the named subset. IMO we should either name the complete permitted set (literally any text other than newline), or use the words "for example" to clarify we are naming a proper subset.

Author: I have updated the PR with more liberal wording. You are sure to have improvements.

to03. Without naming them all explicitly, they appear in the illustrative

>       syntax in subsequent sections.

Reviewer: Unclear what "them" and "they" refer to here.

I think we generally want each rule to stand-alone grammatically, and possibly refer explicitly to other rules when necessary.

> to05. Directive names are recognized at the beginning of a line.

Reviewer: No idea what this rule is trying to say.

Section 2.1 already defined "directive line", but the term "Directive names" has not been defined.

Author: I think "to05" should be moved to the section on whitespace. Im adding a "ws02".

Author: Updated PR to be explicit what we are talking about. Moved to05 to the "whitespace" section.

to07. The '#' character on a directive line may be preceded or followed
      by whitespace.

> to09. The preprocessor is line oriented. To define the end of

Reviewer: ```suggestion
to09. The preprocessor is line-oriented. To define the end of
```
